### DISCUSSION TITLE
I’m just curious . . . where exactly is the magic that makes this work?

### DISCUSSION THREAD

**Discussion by FiveLeavesLeft** [gtxaspec/wz_mini_hacks#46] [General]
I’m just curious . . . where exactly is the magic that makes this work?

I’m just a lowly applications programmer and don’t know anything about embedded linux so I got this book, “Embedded Linux Primer” that I am reading and I am learning about boot loaders. In particular, das u boot which appears to be what wyze uses. I downloaded t31x_uboot from wyze and am looking through it.

 Is it in here somewhere that says, “if something exists on the sd card, then do this”?

Or is it in the linux kernel somewhere that the boot loader launches that looks for something on the sd card?

Am I in the right ballpark?

Thanks for the great project. Are you working on a V2 version?

**Comment by gtxaspec** [comment]
yep.  if you search through the uboot source, I use ```ack```, a nice utility to search all the files in the dir for a string.  

so if you have serial access to the camera, you can look at uboot starting up.  in particular, we see this string printed during boot:  ```factory_t31_ZMC6tiIDQN not found``` or something similar to that

so, in the dir i search for ```ack factory_t31``` which results in that string being in ```common/cmd_sdstart.c```
which is where the magic happens

**Comment by gtxaspec** [comment]
yep, v2 version is in the works, but the problem is that the v2 bootloader doesn't boot stuff off the memory card like the v3 does, unless you compile it yourself, which I did, and bricked the camera.  But i fixed it by removing the flash memory and reprogramming it...still researching.

**Comment by gtxaspec** [comment]
cmd_sdupdate.c has some good stuff too.

**Comment by FiveLeavesLeft** [comment]
Thanks for the leads. Fascinating stuff.

air:t31x_uboot crb$ grep -i -r factory_t31_ZMC6tiIDQN .
./common/cmd_sdstart.c:char *aufiles = "factory_t31_ZMC6tiIDQN";

**Comment by gtxaspec** [comment]
the v2 supported the same thing, just ```factory``` on the v2 uboot, but it's not enabled on the shipping bootloader, or at least not on the one camera i tested.

**Reply by virmaior** [reply]
does this hint that future wyze v3s could have either lack this in uboot or require a different filename? Or is there a reason the T31 needs this?

**Reply by gtxaspec** [reply]
they could in theory, by shipping a camera with a new version of the bootloader, and I doubt they would attempt to update the bootloader over the air, too many chances for the camera to brick if something goes wrong. the filename is set by hualai, the odm developers.
